docs/src/design/proposals/APIRedesign.md

```diff
@@ -0,0 +1,1050 @@
+
+# Goals
+
+## Characteristics of a good API
+
+A good API:
+
+* Is easy to understand.
+* Rewards doing things "the right way."
+* Saves development effort.
+* Is powerful enough to support a broad range of applications.
+* Lends itself to good documentation.
+
+These characteristics can sometimes be at odds with each other, or
+with other concerns. These should typically be viewed as participants
+in trades.
+
+## Evaluating APIs
+
+APIs may be evaluated based on:
+
+* Number of interfaces.
+  * How many application-specific interfaces do I need to know to
+    solve a certain class of problems?
+* Size of interfaces.
+  * How many methods does each interface have?
+* Depth of interfaces.
+  * Specifically, how many methods do I need to call before the return
+    value is of a form that is not specific to the API?
+* Clarity of interfaces.
+  * How much documentation or learning is required before an interface is
+    useful?
+* Consistency of interfaces.
+  * How similar is one interface to an analogous interface?
+* Utility of interfaces.
+  * How much development effort is reduced by utilizing these interfaces,
+    versus accomplishing the same goals with other tools?
+* Power of interfaces.
+  * How much application functionality can I influence with the interfaces
+    that are available to me?
+
+In general, prefer to have a small number of simple, shallow, clear,
+useful, powerful interfaces.
+
+# Developer Feedback
+
+## Developer Intern Feedback
+
+This feedback comes from interns who worked closely with
+Open MCT Web as their primary task over the Summer of 2015.
+
+### Developer Intern 1
+
+Worked on bug fixes in the platform and a plugin for search.
+
+* Initially, it was confusing that many things in files that are in
+  very different locations in the code base refer to each other.
+  * Perhaps explain more the organization strategy behind the
+    different main sections, like "commonUI" vs "core".
+* This may be just me, but there are often long chains of related
+  functions calling each other, and when I had to modify the behavior,
+  I had a hard time remembering to look for the highest level function
+  in the call chain to change. I also sometimes had a hard time finding
+  the connections between the functions. But, that is important because
+  the implementation of the functions along the chain may change later.
+* One very helpful thing that you could add might just be documentation
+  that is not in paragraph format like in the current developer guide.
+  I would just like a list of all the functions and members of each kind
+  of object there is, and descriptions of what they are and how they're
+  used.
+  * Also, the current developer guide pdf's words that are in 'code font',
+   rather than the normal text, are not searchable.
+   (Depending on the pdf viewer.)
+* I do appreciate that there is some example code.
+* I am still slightly confused about what "domainObject" refers to in
```

Comment:
I agree with this. "Domain Object" is not only quite nebulous, but it's also terminology borrowed from the world of software engineering that will be unfamiliar to non software engineers. So long as it stays confined to our code that's ok, but the term should not leak out into any user-facing documentation.